codec-dns-doh/src/main/java/io/netty/handler/codec/dns/doh/DoHResponseDecoder.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright 2020 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.dns.doh;
+
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.MessageToMessageDecoder;
+import io.netty.handler.codec.UnsupportedMessageTypeException;
+import io.netty.handler.codec.dns.DefaultDnsResponse;
+import io.netty.handler.codec.dns.DnsOpCode;
+import io.netty.handler.codec.dns.DnsRecordDecoder;
+import io.netty.handler.codec.dns.DnsResponse;
+import io.netty.handler.codec.dns.DnsResponseCode;
+import io.netty.handler.codec.dns.DnsResponseDecoder;
+import io.netty.handler.codec.http.FullHttpResponse;
+import io.netty.handler.codec.http.HttpObject;
+import io.netty.util.internal.UnstableApi;
+
+import java.net.SocketAddress;
+import java.util.List;
+
+@UnstableApi
+public class DoHResponseDecoder extends MessageToMessageDecoder<HttpObject> {
+
+    private final DnsResponseDecoder<SocketAddress> dnsResponseDecoder;
+
+    /**
+     * Creates a new decoder with {@linkplain DnsRecordDecoder#DEFAULT the default record decoder}.
+     */
+    public DoHResponseDecoder() {
+        this(DnsRecordDecoder.DEFAULT);
+    }
+
+    /**
+     * Creates a new decoder with the specified {@code recordDecoder}
+     */
+    public DoHResponseDecoder(DnsRecordDecoder dnsRecordDecoder) {
+        this.dnsResponseDecoder = new DnsResponseDecoder<SocketAddress>(dnsRecordDecoder) {
+            @Override
+            protected DnsResponse newResponse(SocketAddress sender, SocketAddress recipient, int id,
+                                              DnsOpCode opCode, DnsResponseCode responseCode) {
+                return new DefaultDnsResponse(id, opCode, responseCode);
+            }
+        };
+    }
+
+    @Override
+    protected void decode(ChannelHandlerContext ctx, HttpObject msg, List<Object> out) throws Exception {
+
+        // We don't any other HttpObject except FullHttpResponse
+        if (!(msg instanceof FullHttpResponse)) {
+            throw new UnsupportedMessageTypeException("HttpObject is not FullHttpResponse");
+        }
+
+        FullHttpResponse fullHttpResponse = (FullHttpResponse) msg;
+        out.add(dnsResponseDecoder.decode(ctx.channel().remoteAddress(), ctx.channel().localAddress(),
+                fullHttpResponse.content()));
```

Comment:
Are you taking into account the cache rules at this point? https://tools.ietf.org/html/rfc8484#section-5.1

For example the TTL for records with non zero age?

Reply:
We're not using any cache here. For now, this pull request is a basic implementation of DoH transport with no cache or what-so-ever. I have plans for Cache, QueryDecoder, and ResponseEncoder in the future.

Reply:
I mainly meant that you are dropping the cache headers which can affect the implied TTL in the records.

Reply:
Now I get it. Give me some time, I am on it now.

Reply:
@hyperxpro let me know once resolved

Reply:
@yschimke I read RFC completely. HTTP Header for Caching is useful when we've some sort of caching layer in Netty. For now, we don't have. With that said, users can definitely add their caching own solution at the DnsQuery handler.